daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -95,12 +95,23 @@ private[preprocessing] final class TransactionPreprocessor(
                 commandPreprocessor.unsafePreprocessCreate(create.templateId, create.arg)
               acc.update(newCids, List(create.coid), cmd)
             case exe: Node.NodeExercises[_, Cid] =>
-              val (cmd, newCids) = commandPreprocessor.unsafePreprocessExercise(
-                exe.templateId,
-                exe.targetCoid,
-                exe.choiceId,
-                exe.chosenValue,
-              )
+              import scala.Ordering.Implicits.infixOrderingOps
+              val (cmd, newCids) = exe.key match {
+                case Some(key) if exe.byKey && exe.version >= TransactionVersion.minByKey =>
```

Comment:
@remyhaemmerle-da can you please check if seems sensible? This was (understandably) necessary to fix the conformance tests.

Reply:
> I would prefer we do check for version, i.e. the transaction coder ensures already that exe.byKey is defined only if exe.version >= TransactionVersion.minByKey

That seems fragile. Now it matters if I first roundtrip via serialization & deserialization or whether I get a transaction directly out of the engine and pass it here. I don’t really see a usecase for the second case but I still feel slightly uneasy about breaking it.

Reply:
The validation was original designed to work only on deserialized transaction. I think  we should continue to assume this as: 
- it simplify reasoning about validation as transaction are somehow normalized by the serialization/deserialization process. 
- I do not see use case were we may want to validate non serialized transaction
- The way we are distributing the versioning logic through the code seems to me very fragile, I really want to limit it to some very limited places. You introduce in this PR tow new places.  
- why changing something that has proven to work.

Reply:
I would be fine with writing a transaction (version aware) normalize equivalent to the serialization-deserialization.

Reply:
I was indeed thinking to apply it inside the finish. For canton I see several solution:
 - We add a config flag to normalize or not the `byKey` flag
 - We include the `byKey` flag info in the `Transaction.MetaData`
 - We force Canton to use only LF 1.14 (wish could be somehow possible as there is no Canton production).